File: python/viberpc/http/connection.py
"""HttpConnection — transport connection over HTTP POST."""

from __future__ import annotations

import uuid
import logging
from typing import Any, Callable

import aiohttp

logger = logging.getLogger("viberpc")


class HttpConnection:
    """Wraps a callback URL for bidirectional HTTP POST RPC."""

    def __init__(self, callback_url: str, connection_id: str | None = None) -> None:
        self.callback_url = callback_url
        self.id = connection_id or uuid.uuid4().hex[:12]
        self._open = True
        self._session: aiohttp.ClientSession | None = None

        self.on_message: Callable[[str], Any] | None = None
        self.on_close: Callable[[], Any] | None = None

    @property
    def is_open(self) -> bool:
        return self._open

    async def send(self, raw: str) -> None:
        """POST raw JSON to the peer's callback URL."""
        if not self._open:
            return
        try:
            if self._session is None:
                self._session = aiohttp.ClientSession()
            async with self._session.post(
                self.callback_url,
                data=raw,
                headers={"Content-Type": "application/json"},
            ) as resp:
                if resp.status >= 400:
                    await self._mark_closed()
        except Exception:
            await self._mark_closed()

    async def close(self) -> None:
        await self._mark_closed()
        if self._session:
            await self._session.close()
            self._session = None

    def deliver_message(self, raw: str) -> Any:
        """Deliver a message from the peer (called by HttpServer)."""
        if self.on_message:
            return self.on_message(raw)

    async def _mark_closed(self) -> None:
        if not self._open:
            return
        self._open = False
        if self._session:
            await self._session.close()
            self._session = None
        if self.on_close:
            self.on_close()

File: typescript/src/http/connection.ts
/**
 * HttpConnection — implements ITransportConnection over HTTP POST.
 * Wraps a callback URL (the peer's HTTP endpoint).
 * send(raw) → POST to peer's callback URL.
 */

import type { ITransportConnection } from "../transport.js";

export class HttpConnection implements ITransportConnection {
  onMessage: ((raw: string) => void) | null = null;
  onClose: (() => void) | null = null;

  private _open = true;

  constructor(
    /** The peer's callback URL to POST messages to. */
    public readonly callbackUrl: string,
    /** Optional metadata associated with this connection. */
    public readonly id: string = crypto.randomUUID?.() ?? Math.random().toString(36).slice(2),
  ) {}

  get isOpen(): boolean {
    return this._open;
  }

  send(raw: string): void {
    if (!this._open) return;
    fetch(this.callbackUrl, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: raw,
    }).catch(() => {
      // If POST fails, mark connection as closed
      this._markClosed();
    });
  }

  close(): void {
    this._markClosed();
  }

  /** Deliver a message from the peer (called by HttpServer when it receives a POST). */
  deliverMessage(raw: string): void {
    this.onMessage?.(raw);
  }

  private _markClosed(): void {
    if (!this._open) return;
    this._open = false;
    this.onClose?.();
  }
}
